MCR-2922 Add Neo4J module

bt1cu opened the pull request:
First integration for MyCoRe

[Link to jira](https://mycore.atlassian.net/browse/MCR-2922).

bt1cu committed:
MCR-2922 Add Neo4J module
First integration for MyCoRe

bt1cu committed:
MCR-2922 Add Neo4J

some codacy fixes

bt1cu committed:
MCR-2922 Add Neo4J

fix codacy

bt1cu committed:
Merge branch 'main' into issues/MCR-2922_Add_Neo4J

Hammer1279 committed:
MCR-2922 Add Neo4J
code cleanup, fixes and formatting

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/datamodel/metadata/neo4jparser/MCRNeo4JAbstractDataModelParser.java:
```diff
@@ -0,0 +1,45 @@
+ * along with MyCoRe.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.mycore.datamodel.metadata.neo4jparser;
```
Do not use `org.mycore.datamodel` as package root, because this is already part of `mycore-base`
We must start to keep things separate for the planed switch to Java modules
Use `org.mycore.neo4j` as root

bt1cu replied:
We selected the sub package because it is a directly connection between the parser classes and the datamodel classes. The neo4j datamodel classes use the structure of the MCRMetadataXYZ classes to extract neo4j data. So it seems the best solution to use this name.

rsteph-de replied:
With the planed use of Java modules every JAR file needs a unique package structure.
We've decided, that new modules should already follow the rules (see *mycore-cronjob* or *mycore-acl*).
In your case it could be `org.mycore.mcr.neo4j.datamodel.metadata.neo4jparser`

Hammer1279 committed:
MCR-2922 Add Neo4J
- own java module
- code cleanup, formatting and codacy fixes

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 committed:
MCR-2922 Add Neo4J
replace exception type to non-raw exception

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 committed:
MCR-2922 Add Neo4J
put code in javadoc into code blocks

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 committed:
MCR-2922 Add Neo4J
more code escaped and docs shortened

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 committed:
MCR-2922 Add Neo4J
add summary to everything with author

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 committed:
MCR-2922 Add Neo4J
LIFTing some code around
more summary to be LIFTed about

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

Hammer1279 committed:
MCR-2922 Add Neo4J

JdkObsolete fix, change from LinkedList to ArrayList

Signed-off-by: Tobias Lenhardt <tobias.lenhardt@uni-leipzig.de>

AKluge commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/datamodel/metadata/neo4jutil/Neo4JRelation.java:
```diff
@@ -0,0 +1,74 @@
+
+      public String toFullQuery() {
+            StringBuilder sbQuery = new StringBuilder();
+            Long random = ThreadLocalRandom.current().nextLong(Long.MAX_VALUE);
```
Predictably not relevant since it just used to make the target node reference unique for the query

Hammer1279 committed:
MCR-2922 Add Neo4J
change dir path and remove unused property

rsteph-de commented on mycore-neo4j/src/main/resources/META-INF/web-fragment.xml:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<web-fragment version="5.0" xmlns="https://jakarta.ee/xml/ns/jakartaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Version + Schemalocation auf 6.0, kein ID-Attribut
dann passt das zu den anderen web-fragmenten.
``` 
<web-fragment version="6.0" xmlns="https://jakarta.ee/xml/ns/jakartaee"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/web-fragment_6_0.xsd"
  metadata-complete="true">
```

bt1cu replied:
Version 6 hatte bei mir mal Probleme gemacht, 5 läuft.

rsteph-de commented on mycore-neo4j/src/main/resources/META-INF/web-fragment.xml:
```diff
@@ -0,0 +1,16 @@
+<web-fragment version="5.0" xmlns="https://jakarta.ee/xml/ns/jakartaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/web-fragment_5_0.xsd"
+              metadata-complete="true" id="mymss_common">
+
```
Add name and ordering:
```
  <name>mycore_neo4j</name>
  <ordering>
    <after>
      <name>mycore</name>
    </after>
  </ordering>
```

rsteph-de commented on mycore-neo4j/README.md:
```diff
@@ -0,0 +1,229 @@
+
+1. Gehen Sie zur offiziellen Neo4j-Website: [https://neo4j.com/](https://neo4j.com/download-center/?ref=subscription#community)
+
+2. Klicken Sie auf den "pricing" um zur Downloadseite zu gelangen.
```
... den Menüpunkt "Pricing"

AKluge replied:
done

rsteph-de commented on mycore-neo4j/README.md:
```diff
@@ -0,0 +1,229 @@
+
+# Neo4J MCR-Modul
+
+Das Neo4J MCR-Modul stellt eine Zwischentechnologie, für die Verbindung von Neo4j mit dem MyCore Framework dar.
```
"Brückentechnologie" oder "... stellt eine Verbindung von Neo4J mit dem MyCoRe Framework her." (MyCoRe mit großem "R")

AKluge replied:
done

rsteph-de commented on mycore-neo4j/pom.xml:
```diff
@@ -0,0 +1,120 @@
+  <artifactId>mycore-neo4j</artifactId>
+  <name>MyCoRe Neo4J</name>
+  <properties>
+    <manifest.priority>55</manifest.priority>
```
`55` ist schon für mycore-orcid vergeben -> `57` verwenden

bt1cu replied:
okay

AKluge replied:
done

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/datamodel/metadata/neo4jutil/Neo4JRelation.java:
```diff
@@ -0,0 +1,74 @@
+      public String toAppendQuery() {
+            StringBuilder sbQuery = new StringBuilder();
+            Long random = ThreadLocalRandom.current().nextLong(Long.MAX_VALUE);
+            sbQuery.append("MERGE (b").append(random).append(" {id: '").append(targetNodeID).append("'})\n")
```
Maybe the queries become more readable with text blocks (https://www.baeldung.com/java-multiline-string#text-blocks) and use of String.format() with position arguments:
```
// Specifying argument positions. %1$ is for the first argument and %2$ is for the second argument */
String fstr = String.format("My String is: %1$s, %1$s and %2$s", "cool string", "88");
//fstr = "My String is: cool string, cool string and 88"
```
}
or use Neo4J-Cypher-DSL https://neo4j-contrib.github.io/cypher-dsl/2023.4.0/

AKluge replied:
Hab eine Java Doc Beschreibung zur besseren Erklärung hinzugefügt, da sich schon weiter unten gegen die Verwendung der Cypher-DSL ausgesprochen wurde. Zudem möchte ich gerne bei der Konstanten Verwendung des Stringbuilders bleiben.

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/frontend/cli/MCRNeo4JCommands.java:
```diff
@@ -0,0 +1,156 @@
+   private static final String UTIL_CLASS_PROPERTY = "MCR.Neo4J.Utils.Class";
+   private static final Logger LOGGER = LogManager.getLogger(MCRNeo4JCommands.class.getName());
+
+   @MCRCommand(syntax = "clean neo4j of id {0}", help = "clean the Neo4J database for an ID {0}", order = 10)
```
Ich hätte hier `clean neo4j for id` als Kommando besser gefunden.
In anderen Kommandos haben wir sowohl "of id" als auch "for id" verwendet.

bt1cu replied:
... ist mir auch egal, nehmen wir for :)

AKluge replied:
soll ich hier generell alle of durch for tauschen? oder nur bei den beiden Methoden `clean neo4j of id {0}` und `synchronize neo4j of id {0}` ?

rsteph-de replied:
Ich würde alles ersetzen `for id`, `for type` ,...

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/index/MCRNeo4JIndexEventHandler.java:
```diff
@@ -0,0 +1,220 @@
+package org.mycore.mcr.neo4j.index;
+
+import static org.mycore.mcr.neo4j.datamodel.metadata.neo4jutil.MCRNeo4JConstants.NEO4J_CONFIG_PREFIX;
+import static org.mycore.mcr.neo4j.utils.MCRNeo4JDatabaseDriver.commitWriteOnlyQuery;
```
Ich mag keine statischen Imports von Methoden. Da man beim Aufruf nicht erkennt, wo die Methode herkommt: eigene Klasse, Vererbung, statischer Import, ...

bt1cu replied:
okay

AKluge replied:
hab die statischen imports entfernt

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/index/MCRNeo4JIndexEventHandler.java:
```diff
@@ -0,0 +1,220 @@
+   }
+
+   @Override
+   protected synchronized void handleObjectCreated(MCREvent evt, MCRObject obj) {
```
Ist das `synchronized` hier notwendig, wenn die im Body aufgerufene Methode schon synchronized ist?
In den anderen Eventhandler sind die Methoden auch nicht synchronized.

AKluge replied:
wurde entfernt

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/utils/MCRNeo4JDatabaseDriver.java:
```diff
@@ -0,0 +1,381 @@
+        this.url = url;
+        this.user = user;
+        this.password = password;
+        this.createConnection();
```
nur in diesem Konstruktor wird `createConnection()` aufgerufen?

AKluge replied:
der Constructor wurde entfernt

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/utils/MCRNeo4JDatabaseDriver.java:
```diff
@@ -0,0 +1,381 @@
+ *
+ * @author Andreas Kluge
+ */
+public class MCRNeo4JDatabaseDriver {
```
Was ist das für eine Klasse?
a) normales Objekt - dann kein getInstance() und keine statischen Methoden die mit dem Objekt arbeiten
b) Singleton - dann die Konstruktoren private machen und keine statischen Methoden verwenden
c) Utility-Klasse dann nur statische Methoden und ggf. private Konstruktoren

AKluge replied:
hab b genommen und die query relevanten static Methoden in eine separate  Klasse extrahiert, damit kümmert sich diese Klasse nun wieder ausschließlich um die Verbindung zur Neo4J Datenbank

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/utils/MCRNeo4JDatabaseDriver.java:
```diff
@@ -0,0 +1,381 @@
+
+        List<Neo4JMetaData> neo4JMetaDataList = propertiesMapToJson(startNode.asMap(), lang);
+        String neo4JMetaDataListString = gson.toJson(neo4JMetaDataList);
+
```
JSON per String.format (siehe oben) oder besser per low-level JSON-API (von GSON oder Jackson) erzeugen?
siehe z.B. https://www.tabnine.com/code/java/methods/com.google.gson.JsonObject/%3Cinit%3E?snippet=5ce6e437e594670004069301

AKluge replied:
hab ich so geändert

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/utils/MCRNeo4JUtilsDefault.java:
```diff
@@ -0,0 +1,93 @@
+ * @author Andreas Kluge
+ * @author Jens Kupferschmidt
+ */
+public class MCRNeo4JUtilsDefault implements MCRNeo4JUtilsInterface {
```
Benennung von Klasse und Interface
- `Utils` sind für mich primär Klassen mit statischen Methoden
- `Interface` im Klassennamen ist bei MyCoRe eher unüblich
- Vorschlag: Interface = `MCRNeo4jManager` / Klasse: `MCRNeo4JDefaultManager`

AKluge replied:
hab den Vorschlag so genommen siehe [commit](https://github.com/MyCoRe-Org/mycore/pull/1912/commits/4192aadabc6253d9bd60064c1517376abce22a73)

rsteph-de commented on mycore-neo4j/src/main/resources/config/mycore-neo4j/mycore.properties:
```diff
@@ -0,0 +1,37 @@
+
+# Utils implementation
+  MCR.Neo4J.Utils.Class=org.mycore.mcr.neo4j.utils.MCRNeo4JUtilsDefault
```
Benennung siehe oben

rsteph-de commented on mycore-neo4j/pom.xml:
```diff
@@ -0,0 +1,120 @@
+    <pmd.failOnViolation>false</pmd.failOnViolation>
+  </properties>
+  <build>
+    <plugins>
```
Wir das Datamodel-Plugin hier benötigt?

bt1cu replied:
ein Datenmodell wird nicht generiert

AKluge replied:
da kein Datenmodel generiert wird, habe ich das entfernt

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/datamodel/metadata/neo4jparser/MCRNeo4JParser.java:
```diff
@@ -0,0 +1,213 @@
+                     .map(value -> StringUtils.replace(value, "'", ""))
+                     .map(value -> "'" + value + "'")
+                     .toList();
+                  sbNode.append(", a.").append(key).append('=').append(cleaned);
```
Liste an StringBuffer anfügen ist abhängig von toString() Implementierung
besser String.join() verwenden oder den String gleich aus dem Stream generieren:
`Collectors.joining(delimiter, prefix, suffix)`
https://www.techiedelight.com/collectors-joining-method-java#Collectors_joining

AKluge replied:
also da wir an der Stelle einen Liste brauchen, wäre die Lösung

```
final String cleaned = text.stream()
                     .map(value -> StringUtils.replace(value, "'", ""))
                     .map(value -> "'" + value + "'")
                     .collect(Collectors.joining(", ", "[", "]"));
```

da die queryteil dann bsp so aus sieht a.profession_de = ['beruf_a', 'beruf_b']

rsteph-de replied:
Es könnte sogar noch kürzer gehen:
``` java
final String cleaned = text.stream()
                     .map(value -> StringUtils.replace(value, "'", ""))
                     .collect(Collectors.joining("', '", "['", "']"));
```

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/datamodel/metadata/neo4jparser/MCRNeo4JParser.java:
```diff
@@ -0,0 +1,213 @@
+
+      String sourceNodeInformation = parseSourceNodeInformation(mcrObject, xpf, attributes, xml);
+
+      StringBuilder sbQuery = new StringBuilder();
```
Überlegt mal, ob anstelle dieser "wilden" String-Verkettung der Code mit der Neo4J-Cypher-DSL nicht lesbarer wird:
https://neo4j-contrib.github.io/cypher-dsl/2023.4.0/

bt1cu replied:
Für mich sind String-Verkettungen so besser lesbar als ein magisches Tool. Auch wäre wieder eine Lib mehr im Spiel, was es für mich nicht besser macht. Weniger Referenzen und einfache, wenn auch Lindwurm-artige Konstrukte würde ich einer weiteren Lib vorziehen.

rsteph-de replied:
Ich tue mich mit den "Lindwurm-Strings" schwerer - gerade bei den ganzen Klammern und Anführungsszeichen.
Aber das Abhängigkeits-Argument kann ich nachvollziehen und hat "Gewicht". 
Für mich ist es O.K., wenn ihr auf die Cypher-DSL verzichtet.

toKrause replied:
Zu bedenken ist auch, dass String-konkatenierte Queries ein Einfallstor für Injection-Attacken sein können. Das ist für mich eigentlich der wichtigere Grund für den Einsatz eines ordentlichen Query-Builders, auch wenn ich nicht genug über Neo4j weiß um das Risiko abschätzen zu können. Insofern würde ich doch vorschlagen diesen  vielleicht noch einmal zu besprochen.

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/proxy/MCRNeo4JProxyServlet.java:
```diff
@@ -0,0 +1,215 @@
+      final String id = request.getParameter("id");
+      final String limit = request.getParameter("limit");
+      final String q = request.getParameter("q");
+      final StringBuilder queryStringBuilder = new StringBuilder();
```
Neo4J Cypher-DSL?

bt1cu replied:
siehe oben

AKluge committed:
MCR-2922 Add Neo4J
fix review comment for README.md and pom.xml

AKluge committed:
MCR-2922 Add Neo4J
make MCRNeo4JDatabaseDriver singleton and it only manages the connection with the neo4j Database
extract static methods to own class

AKluge committed:
MCR-2922 Add Neo4J
remove synchronized

AKluge committed:
MCR-2922 Add Neo4J
shorter stream

AKluge committed:
MCR-2922 Add Neo4J
rename MCRNeo4JUtilsDefault to MCRNeo4JDefaultManager

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/utils/MCRNeo4JQueryRunner.java:
```diff
@@ -0,0 +1,236 @@
+            return tx.run(query);
+         });
+      }
+      driver.session().close();
```
close() löschen - durch den try-with-Block darüber ist die Session schon geschlossen.
weitere Code-Stellen suchen und korrigieren

AKluge replied:
hab ich entfernt

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/frontend/cli/MCRNeo4JCommands.java:
```diff
@@ -31,7 +31,7 @@
 import org.mycore.mcr.neo4j.utils.MCRNeo4JDatabaseDriver;
 import org.mycore.mcr.neo4j.utils.MCRNeo4JQueryRunner;
-import org.mycore.mcr.neo4j.utils.MCRNeo4JUtilsInterface;
+import org.mycore.mcr.neo4j.utils.MCRNeo4JManager;
```
Dann sollte konsequenter Weise auch noch das Property angepasst werden:
`UTIL_CLASS_PROPERTY = "MCR.Neo4J.Utils.Class";`

AKluge replied:
NEO4J_MANAGER_CLASS_PROPERTY = "MCR.Neo4J.Manager.Class"; so ?

AKluge committed:
MCR-2922 Add Neo4J
remove unnecessary code

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/frontend/cli/MCRNeo4JCommands.java:
```diff
@@ -0,0 +1,156 @@
+   }
+
+   @MCRCommand(syntax = "test neo4j of id {0}",
+      help = "synchronize metadata to the Neo4J database for MCRID {0}",
```
is this the proper help message?

AKluge replied:
hab die Nachricht geändert

AKluge committed:
MCR-2922 Add Neo4J
change Property

AKluge committed:
MCR-2922 Add Neo4J
update cli runner commands from of to for and update help message for test case

AKluge committed:
MCR-2922 Add Neo4J
some codacy fixes

AKluge committed:
MCR-2922 Add Neo4J
add name and ordering to web-fragment.xml

AKluge committed:
MCR-2922 Add Neo4J
add java doc to clearfy sting building

Hammer1279 commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/frontend/cli/MCRNeo4JCommands.java:
```diff
@@ -0,0 +1,155 @@
+      LOGGER.info("Neo4j connected {}", connected);
+   }
+
+   @MCRCommand(syntax = "test neo4j of id {0}",
```
```suggestion
   @MCRCommand(syntax = "test neo4j for id {0}",
```
All commands use "for" instead of "of" now except for this test command, is this intended?

AKluge committed:
MCR-2922 Add Neo4J
fix typo

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/datamodel/metadata/neo4jparser/MCRNeo4JParser.java:
```diff
@@ -0,0 +1,212 @@
+   private final Map<String, MCRNeo4JAbstractDataModelParser> parserMap;
+
+   public MCRNeo4JParser() {
+      Map<String, String> propertiesMap = MCRConfiguration2.getSubPropertiesMap(NEO4J_CONFIG_PREFIX + "ParserClass.");
```
Is the code redundant to map creation code in MCRNeo4JMetaXMLParser?
If yes ...
- move map to a central location,  UtilityClass?
- move and reuse map creation code?

AKluge replied:
Was moved to util class

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/utils/MCRNeo4JQueryRunner.java:
```diff
@@ -0,0 +1,235 @@
+      return null;
+   }
+
+   private static List<Neo4JMetaData> propertiesMapToJson(Map<String, Object> map, String lang) {
```
Why does name end with 'toJSON?'

AKluge replied:
renamed the function and add a little commentary

rsteph-de commented on mycore-neo4j/src/main/resources/docker/docker-compose.yml:
```diff
@@ -0,0 +1,27 @@
+version: '3.5'
```
These dockerfiles go into the mycore-neo4j.jar .
Is this intended? - Otherwise move to /docker  or src/docker or is there a default-location for docker files in Maven-projects?

rsteph-de replied:
Talked to Sebastian, the files should NOT go into the JAR. 
There is no "default" location - please move to root folder or /docker.

AKluge replied:
moved to mycoire-neo4j/docker

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/utils/MCRNeo4JQueryRunner.java:
```diff
@@ -0,0 +1,235 @@
+   }
+
+   private static String nodeToJson(Node startNode, Gson gson, String lang) {
+      String elementId = startNode.elementId();
```
What's happening here
a) create Json with GSON
b) serialize to String with GSON
c) create Neo4JNodeJsonRecord by parsing the JSON-String via ObjectMapper
d) serialize the object again to String via ObjectMapper

Why don't you simply create the Neo4NodeJsonRecord as Java-POJO?

AKluge replied:
cleaned the function by building the POJO

AKluge committed:
MCR-2922 Add Neo4J
cleanUp nodeTo... function

kkrebs commented:
for string query improvements we created ticket https://mycore.atlassian.net/browse/MCR-2965
can you add TODO-comments to all affected queries with the hint to not copy this string solution!

AKluge committed:
MCR-2922 Add Neo4J
move docker files outside of src folder

AKluge committed:
MCR-2922 Add Neo4J
Move Parser Instantiation Map to Util Class

AKluge committed:
Merge branch 'main' into issues/MCR-2922_Add_Neo4J

AKluge committed:
MCR-2922 Add Neo4J
Fix malformed HTML in javadoc

rsteph-de commented on mycore-neo4j/src/main/java/org/mycore/mcr/neo4j/utils/MCRNeo4JQueryRunner.java:
```diff
@@ -0,0 +1,208 @@
+package org.mycore.mcr.neo4j.utils;
```
MyCoRe Header Comment is missing

AKluge replied:
added

AKluge committed:
MCR-2922 Add Neo4J
add missing MyCoRe Header Comment

rsteph-de reviewed (approved):
looks good for me now
